packages/wallet-lib/src/plugins/Workers/TransactionSyncStreamWorker/methods/processChunks.js

```diff
@@ -1,121 +1,134 @@
-        // and here (https://github.com/nodejs/node/issues/38964)
-        await new Promise((resolveCancel) => setImmediate(() => {
-          self.stream.cancel();
-          resolveCancel();
-        }));
+      if (addressesGenerated > 0) {
+        this.reconnectOnNewBlock = true;
       }
     }
   }
+}
 
-  /* Incoming Merkle block handling */
-  const merkleBlockFromResponse = this.constructor
-    .getMerkleBlockFromStreamResponse(dataChunk);
+async function processMerkleBlock(merkleBlock) {
+  // Ignore Merkle Block processing during the incoming sync
+  // because transactions are verified by actual blocks (handleNewBlock)
+  //
+  // Overall, merkleblocks for freshly mined transactions aren't arriving at all during
+  // the continuous TX sync. (Perhaps due to a bug in dapi?)
+  // However, the stream returns them for blocks N where
+  // subscribeToTransactionsWithProofs({ fromBlockHeight: N, count: 0 })
+  //
+  // This condition ignores such merkle blocks
+  if (this.syncIncomingTransactions) {
+    return;
+  }
+
+  const chainStore = this.storage.getDefaultChainStore();
+
+  const txHashesInTheBlock = merkleBlock
+    .hashes.reduce((set, hashHex) => {
+      const hash = Buffer.from(hashHex, 'hex');
+      hash.reverse();
+      set.add(hash.toString('hex'));
+      return set;
+    }, new Set());
+
+  const headerHash = merkleBlock.header.hash;
+  const headerMetadata = chainStore.state.headersMetadata.get(headerHash);
+
+  if (!headerMetadata) {
+    throw new Error('Header metadata was not found during the merkle block processing');
+  }
 
-  if (merkleBlockFromResponse) {
-    // Reverse hashes, as they're little endian in the header
-    const transactionsInHeader = merkleBlockFromResponse.hashes.map((hashHex) => Buffer.from(hashHex, 'hex').reverse().toString('hex'));
-    const transactionsInWallet = [
-      ...self.storage.getChainStore(self.network).state.transactions.keys(),
-    ];
-    const isTruePositive = isAnyIntersection(transactionsInHeader, transactionsInWallet);
-    if (isTruePositive) {
-      self.importBlockHeader(merkleBlockFromResponse.header);
+  const headerHeight = headerMetadata.height;
+  const headerTime = headerMetadata.time;
+
+  logger.silly(
+    `[TransactionSyncStreamWorker] processing merkle block ${headerHash} at height ${headerHeight}`,
+    { walletId: this.walletId },
+  );
+
+  if (!headerTime || !headerHeight) {
+    throw new Error(`Invalid header metadata: Time: ${headerTime}, Height: ${headerHeight}`);
+  }
+
+  const metadata = {
+    blockHash: headerHash,
+    height: headerHeight,
+    time: new Date(headerTime * 1e3),
+    instantLocked: false, // TBD
+    chainLocked: false, // TBD
+  };
+
+  const transactionsWithMetadata = [];
+  Object.keys(this.transactionsToVerify).forEach((hash) => {
+    const tx = this.transactionsToVerify[hash];
+    if (!txHashesInTheBlock.has(hash)) {
+      throw new Error(`Transaction ${hash} was not found in merkle block ${headerHash}`);
```

Comment:
I think we don't need to throw the error here, transaction might be confirmed in a couple of blocks in the future

Reply:
Obsolete. Refactored TX sync logic https://github.com/dashpay/platform/blob/feat/integrate-with-block-headers-stream/packages/wallet-lib/src/plugins/Workers/TransactionsSyncWorker